File: elixir/1415_get_happy_string.exs
# https://leetcode.com/problems/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/
# 19/02/2025

# 193 ms, 72.8 MB
# defmodule Solution do
#   @spec get_happy_string(n :: integer, k :: integer) :: String.t()
#   def get_happy_string(n, k) do
#     generate_happy_strings(n)
#     |> Enum.at(k - 1, "")
#   end

#   defp generate_happy_strings(n) do
#     generate_happy_strings(["a", "b", "c"], n)
#   end

#   defp generate_happy_strings(strings, n) do
#     if strings |> hd() |> String.length() == n do
#       strings
#     else
#       strings
#       |> generate_new_strings([])
#       |> generate_happy_strings(n)
#     end
#   end

#   defp generate_new_strings([], new_strings), do: Enum.reverse(new_strings)

#   defp generate_new_strings([curr_string | rest_old_strings], new_strings) do
#     case String.last(curr_string) do
#       "a" ->
#         s1 = curr_string <> "b"
#         s2 = curr_string <> "c"
#         generate_new_strings(rest_old_strings, [s2, s1 | new_strings])

#       "b" ->
#         s1 = curr_string <> "a"
#         s2 = curr_string <> "c"
#         generate_new_strings(rest_old_strings, [s2, s1 | new_strings])

#       "c" ->
#         s1 = curr_string <> "a"
#         s2 = curr_string <> "b"
#         generate_new_strings(rest_old_strings, [s2, s1 | new_strings])
#     end
#   end
# end

# 1 ms, 72.4 MB
import Bitwise

defmodule Solution do
  @spec get_happy_string(n :: integer, k :: integer) :: String.t()
  def get_happy_string(n, k) do
    total = 3 * (1 <<< (n - 1))

    if k > total or k <= 0 do
      ""
    else
      build_happy_string(n, k - 1, [], nil)
    end
  end

  defp build_happy_string(n, k, chars, prev_char) do
    current_length = length(chars)

    if current_length == n do
      chars
      |> Enum.reverse()
      |> List.to_string()
    else
      remaining = n - current_length - 1
      possible_chars = possible_chars(prev_char)
      {char, new_k} = find_char(possible_chars, remaining, k)

      if char == nil do
        ""
      else
        build_happy_string(n, new_k, [char | chars], char)
      end
    end
  end

  defp possible_chars(nil), do: [?a, ?b, ?c]

  defp possible_chars(prev_char) do
    case prev_char do
      ?a -> [?b, ?c]
      ?b -> [?a, ?c]
      ?c -> [?a, ?b]
    end
  end

  defp find_char([], _remaining, _k), do: {nil, 0}

  defp find_char([char | rest], remaining, k) do
    count = 1 <<< remaining

    if k < count do
      {char, k}
    else
      find_char(rest, remaining, k - count)
    end
  end
end

defmodule Main do
  def main do
    test_cases = [
      {1, 3, "c"},
      {1, 4, ""},
      {3, 9, "cab"}
    ]

    test_cases
    |> Enum.each(fn {n, k, expected} ->
      result = Solution.get_happy_string(n, k)
      IO.puts("Input: #{inspect({n, k})}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/1415_get_happy_string.rs
// https://leetcode.com/problems/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/
// 19/02/2025

impl Solution {
    pub fn get_happy_string(n: i32, k: i32) -> String {
        let (n, k) = (n as usize, k as usize);
        let mut happy_strings = vec![vec![b'a'], vec![b'b'], vec![b'c']];
        Self::generate_happy_strings(&mut happy_strings, n);
        String::from_utf8(happy_strings.get(k - 1).unwrap_or(&vec![]).to_owned()).unwrap()
    }

    fn generate_happy_strings(strings: &mut Vec<Vec<u8>>, n: usize) {
        if strings[0].len() == n {
            return;
        }

        let mut new_strings: Vec<Vec<u8>> = Vec::with_capacity(strings.len() * 2);
        for s in strings.iter() {
            let (mut s1, mut s2) = (s.clone(), s.clone());
            match s.last() {
                Some(b'a') => {
                    (s1.push(b'b'), s2.push(b'c'));
                }
                Some(b'b') => {
                    (s1.push(b'a'), s2.push(b'c'));
                }
                Some(b'c') => {
                    (s1.push(b'a'), s2.push(b'b'));
                }
                _ => unreachable!(),
            }
            (new_strings.push(s1), new_strings.push(s2));
        }
        *strings = new_strings;
        Self::generate_happy_strings(strings, n);
    }
}

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case(1, 3, "c".to_string())]
    #[case(1, 4, "".to_string())]
    #[case(3, 9, "cab".to_string())]
    fn test(#[case] n: i32, #[case] k: i32, #[case] expected: String) {
        let result = Solution::get_happy_string(n, k);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        (1, 3, "c".to_string()),
        (1, 4, "".to_string()),
        (3, 9, "cab".to_string()),
    ];

    for (n, k, expected) in test_cases {
        let result = Solution::get_happy_string(n, k);
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            (n, k),
            expected,
            result
        );
        assert_eq!(result, expected);
    }
}
